presto-main/src/main/java/com/facebook/presto/operator/spiller/BinaryFileSpiller.java

```diff
@@ -0,0 +1,152 @@
+import java.util.stream.Stream;
+
+import static com.facebook.presto.spi.StandardErrorCode.INTERNAL_ERROR;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+public class BinaryFileSpiller
+        implements Spiller
+{
+    private static final Path SPILL_PATH = Paths.get("/tmp/spills");
+
+    static {
+        SPILL_PATH.toFile().mkdirs();
+    }
+
+    private final Path targetDirectory;
+    private final Closer closer = Closer.create();
+    private final BlockEncodingSerde blockEncodingSerde;
+
+    private int spillsCount = 0;
+    private final ListeningExecutorService executor;
+
+    public BinaryFileSpiller(BlockEncodingSerde blockEncodingSerde, ListeningExecutorService executor)
+    {
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.executor = requireNonNull(executor, "executor is null");
+        try {
+            this.targetDirectory = Files.createTempDirectory(SPILL_PATH, "presto-spill");
+        }
+        catch (IOException e) {
+            throw new PrestoException(INTERNAL_ERROR, "Failed to create spill directory", e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<?> spill(Iterator<Page> pageIterator)
+    {
+        Path spillPath = getPath(spillsCount++);
+
+        return MoreFutures.toCompletableFuture(executor.submit(
+                () -> {
+                    try {
+                        writePages(pageIterator, spillPath);
+                    }
+                    catch (IOException e) {
+                        throw new PrestoException(INTERNAL_ERROR, "Failed to spill pages", e);
+                    }
+                }
+        ));
+    }
+
+    private void writePages(Iterator<Page> pageIterator, Path spillPath)
+            throws IOException
+    {
+        try (OutputStream output = new BufferedOutputStream(new FileOutputStream(spillPath.toFile()))) {
+            PagesSerde.writePages(blockEncodingSerde, new OutputStreamSliceOutput(output), pageIterator);
+        }
+        catch (RuntimeIOException | IOException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    @Override
+    public List<Iterator<Page>> getSpills()
+    {
+        try {
+            ImmutableList.Builder<Iterator<Page>> spills = ImmutableList.builder();
+            for (int spillNumber = 0; spillNumber < spillsCount; spillNumber++) {
+                Path spillPath = getPath(spillNumber);
+                spills.add(readPages(spillPath));
+            }
+
+            return spills.build();
+        }
+        catch (IOException e) {
+            throw new PrestoException(INTERNAL_ERROR, "Failed to read spilled pages", e);
+        }
+    }
+
+    private Iterator<Page> readPages(Path spillPath)
+            throws IOException
+    {
+        InputStream input = new BufferedInputStream(new FileInputStream(spillPath.toFile()));
```

Comment:
does this stream need to be additionally buffered? It is already buffered by operating system.